builds.go

```diff
@@ -55,156 +39,104 @@ type BuildIDParts struct {
 )
 
-type cacheResolveType struct {
-	Resolved *registry.ResolvedObject             `json:"resolved"`
-	Arches   map[string][]registry.ResolvedObject `json:"arches"`
-}
-
-func resolveRemoteArch(ctx context.Context, img string, arch string, diskCacheForSure bool) (*RemoteResolvedFull, error) {
-	cacheFunc, wasCached := cacheResolve.LoadOrStore(img, sync.OnceValues(func() (*cacheResolveType, error) {
-		var (
-			ret = cacheResolveType{}
-			err error
-			individualLookupLimiter = rate.NewLimiter(rate.Every(time.Second), 2) // only do each image lookup at most once per second
-		)
-
-		shouldRetry := func(err error) bool {
-			if err == nil {
-				return false
-			}
-			var ret bool
-			var statusErr *c8derrs.ErrUnexpectedStatus
-			if errors.As(err, &statusErr) {
-				// "Too Many Requests" (new-style containerd error)
-				ret = statusErr.StatusCode == 429
-			} else {
-				// another (older) flavor of 429; https://github.com/containerd/containerd/tree/v1.6.19/remotes/docker/resolver.go#L302
-				ret = strings.Contains(err.Error(), "429 Too Many Requests")
-			}
-			if ret {
-				for i := registryRateLimiter.Tokens(); i > 0; i-- {
-					// just eat all available tokens and starve out the rate limiter
-					_ = registryRateLimiter.Allow()
-				}
-			}
-			return ret
-		}
-
-		for {
-			if err := individualLookupLimiter.Wait(ctx); err != nil {
-				return nil, err
-			}
-			if err := registryRateLimiter.Wait(ctx); err != nil {
-				return nil, err
-			}
-
-			ret.Resolved, err = registry.Resolve(ctx, img)
-			if c8derrdefs.IsNotFound(err) {
-				return nil, nil
-			} else if shouldRetry(err) {
-				fmt.Fprintf(os.Stderr, "warning: lookup %q errored (%q); will retry...\n", img, err)
-				continue
-			} else if err != nil {
-				return nil, err
-			}
-
-			break
-		}
-
-		for {
-			if err := individualLookupLimiter.Wait(ctx); err != nil {
-				return nil, err
-			}
-			if err := registryRateLimiter.Wait(ctx); err != nil {
-				return nil, err
-			}
-
-			// TODO more efficient lookup of single architecture? (probably doesn't matter much, and then we have to have two independent caches)
-			ret.Arches, err = ret.Resolved.Architectures(ctx)
-			if shouldRetry(err) {
-				fmt.Fprintf(os.Stderr, "warning: lookup arches for %q errored (%q); will retry...\n", img, err)
-				continue
-			} else if err != nil {
-				return nil, err
-			}
-
-			break
-		}
+func resolveIndex(ctx context.Context, img string, diskCacheForSure bool) (*ocispec.Index, error) {
+	ref, err := registry.ParseRefNormalized(img)
+	if err != nil {
+		return nil, err
+	}
+	if ref.Digest != "" {
+		// we use "ref" as a cache key, so if we have an explicit digest, ditch any tag data
+		ref.Tag = ""
+	} else if ref.Tag == "" {
+		ref.Tag = "latest"
+	}
+	refString := ref.String()
 
-		return &ret, nil
+	cacheFunc, wasCached := cacheResolve.LoadOrStore(refString, sync.OnceValues(func() (*ocispec.Index, error) {
+		return registry.SynthesizeIndex(ctx, ref)
 	}))
```

Comment:
Technically now that we're transparently caching all <=4MiB registry responses in-memory we don't need this to cache, but it's currently acting as a mutex (and there are some `TODO`s in the cache code related to mutexes that we also need to resolve for this to fully parallelize properly).  There's also enough work being done by `SynthesizeIndex` to make it seem pretty reasonable to leave this as-is for now (otherwise it'll re-parse/re-process the same JSON documents each time it hits the same ref).